Support building pytorch using MKL ILP64 model.

pearu opened the pull request:
This PR enables building pytorch against MKL libraries using the ILP64 model (recall, the currently supported MKL model is LP64).

With the MKL ILP64 model, the 64-bit integer indices of sparse compressed tensors are passed directly to MKL routines which means that the maximum index value can be `2**64-1` instead of the current upper limit of `2**32-1` in the context of blas/lapack usages (recall, currently, 64-bit indices are silently and [unsafely](https://github.com/pytorch/pytorch/issues/102520#issuecomment-1568200510) copied to 32-bit indices before calling the blas/lapack routines).

The PR tackles the following issues: https://github.com/pytorch/pytorch/issues/51720, https://github.com/pytorch/pytorch/issues/100158 .
It illustrates what are the minimal changes to enable MKL ILP64 support in the current main branch.

Usage:
One must define environment variable `MKL_MODEL=ILP64` before building pytorch with MKL ILP64 model support.


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #102613



cc @alexsamardzic @nikitaved @cpuhrsch @amjames @bhosmer @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

pearu committed:
Support building pytorch using MKL ILP64 model.

[ghstack-poisoned]

IvanYashchuk commented on aten/src/ATen/native/BatchLinearAlgebra.h:
```diff
@@ -4,6 +4,16 @@
 #include <ATen/Config.h>
 #include <ATen/native/DispatchStub.h>
 
+#if AT_MKL_ENABLED()
```
ILP64 is not specific to MKL, it's possible to build OpenBLAS with ILP64.

pearu replied:
Right. While slightly OT to this PR, I have inserted
```c++
#ifdef OPENBLAS_USE64BITINT
#define INT_T int64_t
```

However, I suspect that the support for building pytorch with OpenBLAS ILP64 model requires more than that. For instance, detecting the openblas64 library is not implemented in cmake/Modules/FindOpenBLAS.cmake.

foxtran replied:
I wonder why PyTorch does not use standard CMake Modules.  Since CMake 3.22, it has a support of finding the specific LP64/IPL64 models (https://cmake.org/cmake/help/v3.22/module/FindBLAS.html)

IvanYashchuk replied:
For no specific reason. PyTorch maintains its own CMake Modules and somebody should do the update of https://github.com/pytorch/pytorch/blob/main/cmake/Modules/FindBLAS.cmake

IvanYashchuk commented on aten/src/ATen/mkl/Utils.h:
```diff
@@ -15,3 +17,22 @@ static inline MKL_INT mkl_int_cast(int64_t value, const char* varname) {
+
+#define MKL_TENSOR_DATA_PTR(INPUT) reinterpret_cast<MKL_INT*>((INPUT).data_ptr<TORCH_COMPATIPLE_MKL_INT>())
+#define MKL_TENSOR_MUTABLE_DATA_PTR(INPUT) reinterpret_cast<MKL_INT*>((INPUT).mutable_data_ptr<TORCH_COMPATIPLE_MKL_INT>())
+#define MAGMA_TENSOR_DATA_PTR(INPUT) reinterpret_cast<magma_int_t*>((INPUT).data_ptr<TORCH_COMPATIPLE_MKL_INT>())
+#define MAGMA_TENSOR_MUTABLE_DATA_PTR(INPUT) reinterpret_cast<magma_int_t*>((INPUT).mutable_data_ptr<TORCH_COMPATIPLE_MKL_INT>())
```
This file is inside `mkl` folder and it is not an appropriate place for MAGMA macros.

pearu replied:
Moved these macros to aten/src/ATen/native/cuda/linalg/MagmaUtils.h

pearu committed:
Update on "Support building pytorch using MKL ILP64 model."

This PR enables building pytorch against MKL libraries using the ILP64 model (recall, the currently supported MKL model is LP64).

With the MKL ILP64 model, the 64-bit integer indices of sparse compressed tensors are passed directly to MKL routines which means that the maximum index value can be `2**64-1` instead of the current upper limit of `2**32-1` in the context of blas/lapack usages (recall, currently, 64-bit indices are silently and [unsafely](https://github.com/pytorch/pytorch/issues/102520#issuecomment-1568200510) copied to 32-bit indices before calling the blas/lapack routines).

The PR tackles the following issues: https://github.com/pytorch/pytorch/issues/51720, https://github.com/pytorch/pytorch/issues/100158 .
It illustrates what are the minimal changes to enable MKL ILP64 support in the current main branch.

Usage:
One must define environment variable `MKL_MODEL=ILP64` before building pytorch with MKL ILP64 model support.





cc alexsamardzic nikitaved cpuhrsch amjames bhosmer jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

pearu committed:
Update on "Support building pytorch using MKL ILP64 model."

This PR enables building pytorch against MKL libraries using the ILP64 model (recall, the currently supported MKL model is LP64).

With the MKL ILP64 model, the 64-bit integer indices of sparse compressed tensors are passed directly to MKL routines which means that the maximum index value can be `2**64-1` instead of the current upper limit of `2**32-1` in the context of blas/lapack usages (recall, currently, 64-bit indices are silently and [unsafely](https://github.com/pytorch/pytorch/issues/102520#issuecomment-1568200510) copied to 32-bit indices before calling the blas/lapack routines).

The PR tackles the following issues: https://github.com/pytorch/pytorch/issues/51720, https://github.com/pytorch/pytorch/issues/100158 .
It illustrates what are the minimal changes to enable MKL ILP64 support in the current main branch.

Usage:
One must define environment variable `MKL_MODEL=ILP64` before building pytorch with MKL ILP64 model support.





cc alexsamardzic nikitaved cpuhrsch amjames bhosmer jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

pearu commented:
@pytorchbot rebase
